ui/lib/kv/addon/components/kv-patch-editor.js

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+import { A } from '@ember/array';
+import { hasWhitespace, isNonString } from 'vault/utils/validators';
+
+const WHITESPACE_WARNING =
+  "This key contains whitespace. If this is desired, you'll need to encode it with %20 in API requests.";
+
+const NON_STRING_WARNING =
+  'This value will be saved as a string. If you need to save a non-string value, please use the JSON editor.';
+
+class Kv {
```

Comment:
This class tracks the state and values of each key value object that's been added to the tracked array. the empty row in the form (always the last row) is the only row that tracked variables in the component class keep track of

<img width="1104" alt="Screenshot 2024-08-12 at 6 38 46 PM" src="https://github.com/user-attachments/assets/571812dc-49e4-4a55-9db3-ba4b64cc4424">